File: util/sntable_dump.pl
#!/usr/bin/perl
#
# Created April 2013 by R.Kessler
# Utility to examine and dump analysis info from any SNTABLE
# in hbook (ntuple) or root (tree) format.
# The basic functions here are
#  - print list of all available variables in SNTABLE
#  - dump subset of variables for each event into text file.
#
# This utility is aimed for those who prefer NOT to use 
# hbook or root, and prefer to extract analysis variables
# into text files.
#
#
# USAGE:
#   sntable_dump.pl <inFile>
#      (list tables in <inFile>: hbook or root file)
#
#   sntable_dump.pl <inFile>  <tableID>
#      (list variable names in table, then exit)
#
#   sntable_dump.pl <inFile>  <tableID>  --v <list of variables>
#      (dump values for each table entry)
#
#   sntable_dump.pl <inFile>  <tableID>  --v <list of variables> --o <outFile>
#     (overwrite default outfile in ASCII/TEXT format)
#
#   sntable_dump.pl <inFile>  <tableID>  --v <list of variables> NOHEADER
#     (out file has no header info and no 'SN:' key)
#
#   sntable_dump.pl <inFile>  <tableID>  --v <vlist>  --a <textFile>
#     (append fitres text file with vlist variables)
#
#   sntable_dump.pl <inFile>  <tableID>  --outlier <Nsig1> <Nsig2>
#    (dump epochs for data-fit outlier between Nsig1 and Nsig2 sigma)
#
#   sntable_dump.pl <inFile>  <tableID>  --outlier_sim <Nsig1> <Nsig2>
#    (dump epochs for data-sim outlier between Nsig1 and Nsig2 sigma)
#
#   sntable_dump.pl <inFile>  <tableID>  --outlier <Nsig1> <Nsig2> \
#          --format IGNORE
#    (dump outliers in IGNORE format for SNANA data version)
#
#   sntable_dump.pl <inFile>  <tableID>  OBS
#     (dump each observation, mainly for fluxerrmap)
#
#  Note that --[v,o,a] and -[V,O,A] are accpeted (both lower,upper case)
#
#
#      HISTORY
#
# Apr 26 2014: allow HBOOK/hbook extenstion along with HIS/.his
#
# Aug 2014: new argument  '--outlier <n1> <n2>'
#
# Oct 28, 2014: .txt -> .text for output.
#
# Jan 10 2015: use H or R argument for combine_fitres so that output
#              table format (root or hbook) matches input table format.
#
# May 10 2016: 
#   for HBOOK files, allow root-string table names SNANA & FITRES.
#
# Feb 22 2018: new  OBS argument to dump observations for fluxerrmap
#
# ------------------------------------------

use FindBin qw($Bin);
use lib "$Bin" ;
use sntools ;
use strict ;

# ----------- global declarations ------------

my $JOBNAME = "sntable_dump.exe" ;
my $IFILETYPE_HBOOK    = 1 ; 
my $IFILETYPE_ROOT     = 2 ; 
my @COMINE_FITRES_ARG  = ( "?", "H", "R" );
my @VALID_SUFFIX_HBOOK = ( ".HIS", ".HBOOK", ".TUP"  ) ;
my @VALID_SUFFIX_ROOT  = ( ".ROOT" ) ;

# inputs
my ($TABLE_INFILE, $TABLE_ID, @TABLE_VARLIST, $NVARLIST);
my ($OUTFILE_TEXT, $OPT_HEAD, $TEXTFILE_APPEND);
my (@OUTLIER_NSIGMA, $KEY_OUTLIER, $FORMAT_OUTFILE, $OPTOBS );


# misc. globals
my (@MSGERR, $CMD_DUMP, $IFILETYPE );

# -----------  function declarations -----------
sub parse_args ;
sub insert_ccid_varlist ;
sub get_fileType ;
sub check_TABLE_ID_ROOT ;
sub check_TABLE_ID_HBOOK ;
sub make_dump_command ;
sub run_dump_command ;
sub textfile_append ;

# ================ BEGIN MAIN ===============

# parse command-line args
&parse_args() ;


# make sure CCID is 1st element of VARLIST
&insert_ccid_varlist();  


# determine root or hbook file type
&get_fileType()  

# check table ID
&check_TABLE_ID_ROOT();
&check_TABLE_ID_HBOOK();

# construct C-program command 
&make_dump_command();  


# run C program
if ( &run_dump_command() != 0 ) { die "" ; }


# check option to append fitres file with table-extracted variables
&textfile_append();


# ================ END MAIN =================

sub parse_args {

    my ($NARG,  $i, $VFLAG, $USE, $NBAD, $KEY );
    $NARG = scalar(@ARGV); 

    if ( $NARG < 1 ) {
	$MSGERR[0] = "Must give table-File as arguments";
	sntools::FATAL_ERROR(@MSGERR); 
    }

    # init defaults
    $TABLE_INFILE    = "NULL" ;
    $TABLE_ID        = "" ;
    @TABLE_VARLIST   = ( );
    $OUTFILE_TEXT    = "" ;
    $TEXTFILE_APPEND = "" ;
    $OPT_HEAD        = "" ; # default (blank) is to add header
    $VFLAG = 0 ;
    
    @OUTLIER_NSIGMA = ( -9, -9 );
    $KEY_OUTLIER    = "xxx" ;
    $OPTOBS         = 0 ;

    $FORMAT_OUTFILE = "DEFAULT" ; # default is key format

    $NBAD = 0;

    # ------------------------------
    $TABLE_INFILE = $ARGV[0] ;

    if ( $NARG >= 2 ) { $TABLE_ID  = $ARGV[1] ; }
    
    $i = 2;
    while ($i < $NARG) {

	$USE = 0 ;

	if ( $ARGV[$i] eq "NOHEADER" ) 
	{ $USE=1; $OPT_HEAD = "NOHEADER" ; $VFLAG = 0 ; }

	if ( lc($ARGV[$i]) eq '--o' ) 
	{ $USE=1; $i++ ; $OUTFILE_TEXT = $ARGV[$i] ; $VFLAG=0; }

	if ( lc($ARGV[$i]) eq '--a' ) 
	{ $USE=1; $i++; $TEXTFILE_APPEND = $ARGV[$i] ; $VFLAG=0; }

	if ( $VFLAG ) 
	{ $USE=1; @TABLE_VARLIST = (@TABLE_VARLIST , $ARGV[$i]) ; }
	if ( lc($ARGV[$i]) eq '--v'  ) { $USE=1; $VFLAG = 1; }
	
	$KEY = "outlier" ;
	if ( lc($ARGV[$i]) eq "--$KEY" )  { 
	    $USE=1; 
	    $KEY_OUTLIER = "$KEY";
	    $i++ ; $OUTLIER_NSIGMA[0] = $ARGV[$i] ; 
	    $i++ ; $OUTLIER_NSIGMA[1] = $ARGV[$i] ; 
	}

	$KEY = "outlier_sim" ;
	if ( lc($ARGV[$i]) eq "--$KEY" )  { 
	    $USE=1; 
	    $KEY_OUTLIER = "$KEY";
	    $i++ ; $OUTLIER_NSIGMA[0] = $ARGV[$i] ; 
	    $i++ ; $OUTLIER_NSIGMA[1] = $ARGV[$i] ; 
	}


	$KEY = "format" ;
	if ( lc($ARGV[$i]) eq "--$KEY" )  { 
	    $USE = 1 ;
	    $i++ ; $FORMAT_OUTFILE = $ARGV[$i] ;
	}

	if ( lc($ARGV[$i]) eq "obs" )  { 
	    $USE=1; 
	    $OPTOBS = 1;
	}

	if ( $USE == 0 ) {
	    $NBAD++ ;
	    print " ERROR: Unknown argument '$ARGV[$i]' \n";
	}

	$i++ ;
    }

    if ( $NBAD > 0 ) 
    { die "\n FATAL ERROR: $NBAD Unknown args \n"; }



    # if OUTFILE_TEXT is not set, then set to default name using TABLE_ID
    if ( length($OUTFILE_TEXT) == 0 )  { 
	$OUTFILE_TEXT = "sntable_dump_${TABLE_ID}.fitres"; 
	if ( $OUTLIER_NSIGMA[0] >= 0 ) 
	{ $OUTFILE_TEXT = "sntable_dump_${KEY_OUTLIER}.fitres"; }
	if ( $OPTOBS ) 
	{ $OUTFILE_TEXT = "sntable_dump_obs.fitres"; }
    }
    

    $NVARLIST = scalar(@TABLE_VARLIST);

    print " Input table file : $TABLE_INFILE \n";
    print " Input table ID   : $TABLE_ID \n";

    if ( $NVARLIST > 0 ) { 
	print " List of variables to dump: @TABLE_VARLIST \n" ; 
	print " Output text file : $OUTFILE_TEXT \n" ;
    }

    if ( length($TEXTFILE_APPEND) > 0 ) {
	print " Append variables to : $TEXTFILE_APPEND \n";
	$OPT_HEAD = "" ;  # must have default header
    }

    print "\n";

} # end of parse_args


# ===================
sub insert_ccid_varlist {
    
    #
    # Make sure that CCID is first element of variable list.
    # -> first remove CCID or CID from wherever it is.
    # -> add CCID as first element.
    #

    if ( $NVARLIST == 0        ) { return ; }
    if ( $TABLE_ID eq "GLOBAL" ) { return ; }

    # don't add CID for outlier option since CID will be added 
    # internally by sntable_dump.exe (Aug 4 2014)
    if ( $OUTLIER_NSIGMA[0] >= 0 ) { return ; }
    if ( $OPTOBS                 ) { return ; }

    my ($VAR);
    my @TMPLIST = @TABLE_VARLIST ;

    @TABLE_VARLIST = ( "CCID" );

    foreach $VAR (@TMPLIST) {
	if ( $VAR eq "CCID" ) { next; }
	if ( $VAR eq "CID"  ) { next; }
	@TABLE_VARLIST  = ( @TABLE_VARLIST , $VAR ) ;
    }

} # end of insert_ccid_varlist

# ==============================
sub make_dump_command {

    $CMD_DUMP = "$JOBNAME  $TABLE_INFILE  $TABLE_ID  $OPT_HEAD" ;

    if ( $NVARLIST > 0 ) { 
	$CMD_DUMP = "$CMD_DUMP --v @TABLE_VARLIST" ; 
	$CMD_DUMP = "$CMD_DUMP --o $OUTFILE_TEXT" ; 
    }

    if ( $OUTLIER_NSIGMA[0] >= 0 ) {
	$CMD_DUMP = "$CMD_DUMP --$KEY_OUTLIER @OUTLIER_NSIGMA" ;
	$CMD_DUMP = "$CMD_DUMP --o $OUTFILE_TEXT" ; 
    }

    if ( $OPTOBS ) {
	$CMD_DUMP = "$CMD_DUMP  OBS  --o $OUTFILE_TEXT" ; 
    }

    if ( $FORMAT_OUTFILE ne "DEFAULT" ) {
	$CMD_DUMP = "$CMD_DUMP --format $FORMAT_OUTFILE" ; 
    }

    my $SNANA_DIR = $ENV{'SNANA_DIR'};
    print " SNANA_DIR:  $SNANA_DIR \n";
    print " DUMP COMMAND: \n    $CMD_DUMP \n";

} # end of make_dump_command



# ===========================
sub get_fileType {

    # set global IFILETYPE (hbook or root)

    my ($TF, $SUF, $suf);

    $TF = $TABLE_INFILE ;
    $IFILETYPE = -9;

    foreach $SUF ( @VALID_SUFFIX_HBOOK) {
	$suf = lc($SUF); # allow lower case
	if ( index($TF,$SUF)  > 0 ) { $IFILETYPE = $IFILETYPE_HBOOK ; }
	if ( index($TF,$suf)  > 0 ) { $IFILETYPE = $IFILETYPE_HBOOK ; }
    }

    foreach $SUF ( @VALID_SUFFIX_ROOT) {
	$suf = lc($SUF); # allow lower case
	if ( index($TF,$SUF)  > 0 ) { $IFILETYPE = $IFILETYPE_ROOT ; }
	if ( index($TF,$suf)  > 0 ) { $IFILETYPE = $IFILETYPE_ROOT ; }
    }

    if ( $IFILETYPE < 0 ) {
	$MSGERR[0] = "Cannot determine file type (hbook or root)" ;
	$MSGERR[1] = "for $TF ." ;
	$MSGERR[2] = "Allowed table-file extensions are: ";
	$MSGERR[3] = " @VALID_SUFFIX_HBOOK  @VALID_SUFFIX_ROOT \n";
	sntools::FATAL_ERROR(@MSGERR); 
    }


} # end of get_fileType


# ==============================
sub check_TABLE_ID_ROOT {

    # for root file, allow old NTUPLE IDs 7100 or 7788 by
    # translating them into their root names: SNANA or FITRES.

    my ($TID_ORIG, $REPLACE) ;

    if ( $IFILETYPE != $IFILETYPE_ROOT ) { return ; }

    $TID_ORIG = $TABLE_ID ;
    $REPLACE = 0 ;

    if ( $TABLE_ID == 7100 ) 
    { $TABLE_ID = "SNANA" ;  $REPLACE = 1;  }


    if ( $TABLE_ID == 7788 ) 
    { $TABLE_ID = "FITRES" ;  $REPLACE = 1;  }

    if ( $REPLACE ) {
	print "\n";
	print "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n";
	print "    WARNING: \n" ;
	print "\t Replacing hbook NTID=$TID_ORIG  \n";
	print "\t with root table name: $TABLE_ID\n";
	print "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n";
	print "\n";
    }


} # end of check_TABLE_ID_ROOT

# ==============================
sub check_TABLE_ID_HBOOK {

    # for HBOOK file, allow root-string table names FITRES & SNANA.

    my ($TID_ORIG, $REPLACE) ;

    if ( $IFILETYPE != $IFILETYPE_HBOOK ) { return ; }

    $TID_ORIG = $TABLE_ID ;
    $REPLACE = 0 ;

    if ( $TABLE_ID eq "SNANA" ) 
    { $TABLE_ID = 7100 ;  $REPLACE = 1;  }


    if ( $TABLE_ID eq "FITRES" ) 
    { $TABLE_ID = 7788 ;  $REPLACE = 1;  }

    if ( $REPLACE ) {
	print "\n";
	print "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n";
	print "    WARNING: \n" ;
	print "\t Replacing root table=$TID_ORIG  \n";
	print "\t with HBOOK ntuple ID: $TABLE_ID\n";
	print "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n";
	print "\n";
    }


} # end of check_TABLE_ID_HBOOK

# ===========================
sub run_dump_command {

    my ($istat);

    if ( -e $OUTFILE_TEXT ) { qx(rm $OUTFILE_TEXT) ; }

    $istat = system("$CMD_DUMP");

    return $istat ;

} # end of run_dump_command


# ============================
sub textfile_append {

    if ( length($TEXTFILE_APPEND) == 0 ) { return ; }

    my ($CMD, $ARGF, $ARGO, $OUTFILE_PREFIX, $TEXTFILE_PREFIX, $j);

    # use text file prefix to construct output prefix to make
    # sure that it is unique -> avoid clobbering with multiple jobs.

    $j = index($TEXTFILE_APPEND,".");
    $TEXTFILE_PREFIX = substr($TEXTFILE_APPEND,0,$j);
    $OUTFILE_PREFIX  = "sntable_append_${TEXTFILE_PREFIX}" ;

    print "\n Append $TEXTFILE_APPEND with $OUTFILE_TEXT ... \n";
    
    $ARGF = "$TEXTFILE_APPEND $OUTFILE_TEXT $COMINE_FITRES_ARG[$IFILETYPE]" ;
    $ARGO = "--outprefix $OUTFILE_PREFIX" ;
    $CMD  = "combine_fitres.exe $ARGF $ARGO" ;
    
    print "CMD = $CMD \n";
    system("$CMD");

    # ------------------------------
    # insert few lines of commentary at top of appended fitres file.

    my $qq = '"' ;
    my @VARPRINT = @TABLE_VARLIST[1 .. $#TABLE_VARLIST] ;

    my $comment1 = "# Appended variables '@VARPRINT'" ;
    my $comment2 = "# from $TABLE_INFILE  (table $TABLE_ID)";
    my $comment3 = "# ";

    my ($sedcmd, $outFile, $tmpFile);
    $outFile = "${OUTFILE_PREFIX}.text" ;
    $tmpFile = "TEMP_$outFile" ;

    $sedcmd = "sed ";
    $sedcmd = "$sedcmd -e 1i${qq}${comment1}${qq}";
    $sedcmd = "$sedcmd -e 1i${qq}${comment2}${qq}";
    $sedcmd = "$sedcmd -e 1i${qq}${comment3}${qq}";
    qx($sedcmd $outFile > $tmpFile ; mv $tmpFile $outFile);

} # end of  textfile_append

File: util/sntable_dump.py
#!/usr/bin/env python
#
# Created March 2021 by R.Kessler
#   [re-write of perl script sntable_dump.pl]
#
# Utility to examine and dump analysis info from any SNTABLE
# in hbook (ntuple) or root (tree) format.
# The basic functions here are
#  - print list of all available variables in SNTABLE
#  - dump subset of variables for each event into text file.
#  - extract flux outliers w.r.t. sim truth or LC fit curve.
#
# This utility is aimed for those who prefer NOT to use 
# hbook or root, and prefer to extract analysis variables
# into text files.
#
# This python script is just a shell that calls underyling C program
# sntable_dump.exe .
#
# Apr 7 2021
#   + add flush statements
#   + exit if Cprogram aborts
#   + return error code for submit_batch_jobs
#
# Jul 14 2023
#   + call print_elapse_time for VERBOSE mode
#   + for combine_fitres.exe call in append mode, add 't' argument to 
#     write only text output and avoid writing useless ROOT file.
# ===================

import os, sys, argparse, yaml, datetime 
from   argparse import Namespace

VARNAME_CCID = "CCID"
FORMAT_ROOT  = "ROOT"
FORMAT_HBOOK = "HBOOK"

# define maps to correct user mistakes on table names because
# HBOOK takes table number while root takes tree name.
# Can specify SNANA or FITRES table name for any format,
# and the HBOOK map internally translates to integer ID
TABLE_NAME_MAP_HBOOK = { 
    'SNANA'   : '7100',    
    'FITRES'  : '7788',
    'OUTLIER' : '7800'
}
TABLE_NAME_MAP_ROOT  = { 
    '7100'  : 'SNANA',   
    '7788'  : 'FITRES',
    '7800'  : 'OUTLIER'
}

OUTFILE_PREFIX = "sntable_dump"
OUTFILE_SUFFIX = "fitres"

Cprogram_dump    = "sntable_dump.exe"
Cprogram_combine = "combine_fitres.exe"

# ===========
def get_args():
    parser = argparse.ArgumentParser()
    
    msg = "name of table file to examine"
    parser.add_argument("input_table_file", help=msg, nargs="?", default=None)

    msg = "name of table (e.g., SNANA, FITRES, SPEC ...)"
    parser.add_argument("table_name", help=msg, nargs="?", default=None)

    msg = "List possible table names"
    parser.add_argument("-l", "--list_tables", help=msg, action="store_true")

    msg = "comma-sep list of variables to extract"
    parser.add_argument("-v", "--varlist", help=msg, nargs='?', 
                        type=str, default=None)

    msg = "Extract data-fit outliers with Nsig_min < |PULL| < Nsig_max"
    parser.add_argument("--outlier_fit", metavar=('Nsig_min', 'Nsig_max'),
                        help=msg, nargs=2, type=float, default=None)

    msg = "Extract data-sim outliers with Nsig_min < |PULL| < Nsig_max"
    parser.add_argument("--outlier_sim", metavar=('Nsig_min', 'Nsig_max'),
                        help=msg, nargs=2, type=float, default=None)

    msg = "append varlist to this already existing fitres text file"
    parser.add_argument("-a", "--append_file", help=msg, nargs='?', 
                        type=str, default=None)

    msg = "output file name for text formatted table"
    parser.add_argument("-o", "--outfile", help=msg, nargs='?', 
                        type=str, default=None)

    msg = "Print Number of events in table"
    parser.add_argument("-N", "--NEVT", help=msg, action="store_true")

    msg = "dump each observation, for FLUXERR map"
    parser.add_argument("-O", "--OBS", help=msg, action="store_true")

    msg = "Verbose mode"
    parser.add_argument("-V", "--VERBOSE", help=msg, action="store_true")

    args = parser.parse_args()

    return args
    # end get_args

def list_tables():

    print(" Possible table names (2nd arg for this script)")
    print(" from &SNLCINP input SNTABLE_LIST, and used by")
    print(" snana.exe, snlc_fit.exe, psnid.exe programs:")
    for table in TABLE_NAME_MAP_HBOOK:
        print(f"\t{table}")

    sys.exit("\nBye Bye.\n")
    # end list_tables

def insert_ccid_varlist(args):

    # append CCID to varlist, and remove commas

    # bail if CCID is already in varlist
    varlist = args.varlist
    if varlist is None : return None

    if VARNAME_CCID in varlist: return varlist

    varlist_new = f"{VARNAME_CCID},{varlist}"

    varlist_new = varlist_new.replace(',',' ')

    # replace commas with spaces for C program
    return varlist_new

    # end insert_ccid_varlist

def get_file_format(input_table_file):

    Format = None
    
    for suffix in [ 'ROOT', 'root' ]:
        if suffix in input_table_file: Format = FORMAT_ROOT

    for suffix in [ 'HBOOK', 'hbook' ]:
        if suffix in input_table_file: Format = FORMAT_HBOOK

    if Format is None :
        sys.exit(f"\n ERROR: unknown format for " \
                 f"input_table_file = {input_table_file}\n")

    return Format
    # end get_file_format

def check_table_name(table_name_orig,Format):

    table_name_new  = table_name_orig

    if Format == FORMAT_HBOOK:
        if table_name_orig in TABLE_NAME_MAP_HBOOK:
            table_name_new = TABLE_NAME_MAP_HBOOK[table_name_orig]

    if Format == FORMAT_ROOT:
        if table_name_orig in TABLE_NAME_MAP_ROOT:
            table_name_new = TABLE_NAME_MAP_ROOT[table_name_orig]

    return table_name_new

    # end check_table_name

def get_outfile_name(input_args,config):

    outfile    = input_args.outfile
    table_name = config.table_name 
   
    if outfile :  return outfile

    # default name is based on name of table
    outfile = f"{OUTFILE_PREFIX}_{table_name}.{OUTFILE_SUFFIX}"

    if input_args.OBS:
        outfile = f"{OUTFILE_PREFIX}_obs.{OUTFILE_SUFFIX}"

    if input_args.outlier_fit:
        outfile = f"{OUTFILE_PREFIX}_outlier_fit.{OUTFILE_SUFFIX}"

    if input_args.outlier_sim:
        outfile = f"{OUTFILE_PREFIX}_outlier_sim.{OUTFILE_SUFFIX}"

    return outfile

    # end get_outfile_name

def make_command_sntable_dump(input_args,config):

    # construct command for C program

    outfile_flag = False 
    varlist      = config.varlist
    table_file   = input_args.input_table_file
    outlier_fit  = input_args.outlier_fit
    outlier_sim  = input_args.outlier_sim
    table_name   = config.table_name
    outfile      = config.outfile

    command = f"{Cprogram_dump} {table_file} "
    
    if table_name is not None:
        command += f"{table_name} "
    else:
        return command

    if varlist is not None:
        command += f"-v {varlist} "
        outfile_flag = True

    if input_args.NEVT :
        command += f" NEVT "  

    if outlier_fit :
        outlier_list = " ".join(map(str,outlier_fit))
        command += f"--outlier_fit {outlier_list} "
        outfile_flag = True

    if outlier_sim :
        outlier_list = " ".join(map(str,outlier_sim))
        command += f"--outlier_sim {outlier_list} "
        outfile_flag = True

    if input_args.OBS:
        command += "obs "
        outfile_flag = True
        
    if outfile_flag :
        command += f"-o {outfile} "

    return command 
    # end make_command_sntable_dump


def append_fitres(input_args,config):

    # combine args.append file with config.outfile

    input_table_file = input_args.input_table_file
    append_file      = input_args.append_file
    varlist          = (input_args.varlist).replace(',',' ')
    outfile_dump     = config.outfile
    table_name       = config.table_name
    Format           = config.Format
    combine_arg      = Format[0]  # H or R to make HBOOK or ROOT file

    tmp = append_file.split('.')
    outfile_prefix = f"sntable_append_{tmp[0]}"
    outfile        = f"{outfile_prefix}.text"

    cmd = f"{Cprogram_combine} {append_file} {outfile_dump} {combine_arg} " \
          f"-outprefix {outfile_prefix} " \
          f"t"   # <== write only text output; no ROOT or hbook

    os.system(cmd)

    if input_args.VERBOSE:
        print(f" Finished {Cprogram_combine} with append mode.")
        t0 = config.t0
        print_elapse_time(t0)
        sys.stdout.flush()

    # append comments at top of appended fitres file
    comment_list = []
    comment_list.append(f"# Appended variables '{varlist}' ")
    comment_list.append(f"# from {input_table_file}  (table {table_name})")
    comment_list.append(f"# ")

    comment_file = "tmp_comments_append.text"
    with open(comment_file,"wt") as f:
        for comment in comment_list:
            f.write(f"{comment}\n")
    
    # catenate comments and append file
    cmd_cat = f"cat {outfile} >> {comment_file} ; mv {comment_file} {outfile}"
    os.system(cmd_cat)

    if input_args.VERBOSE:
        print(f" Finished adding comments to top of appended file.")
        t0 = config.t0
        print_elapse_time(t0)
        sys.stdout.flush()

    # end append_fitres

def print_elapse_time(t0):
    dt_sec = (datetime.datetime.now() - t0).total_seconds()
    print(f"\t elapse time: {dt_sec} sec")

# =========================
# ======= MAIN ============
# =========================

if __name__ == "__main__":
    
    input_args   = get_args()
    config       = Namespace()

    if input_args.VERBOSE :
        print(f"\n# =============== BEGIN =============== ")
        command = ' '.join(sys.argv)
        print(f" Full python command: \n{command} \n")
        t0 = datetime.datetime.now()
        config.t0 = t0
        sys.stdout.flush()

    if input_args.list_tables: list_tables()

    # insert CCID as first variable
    config.varlist = insert_ccid_varlist(input_args)

    # check file format: root, hbook, ...
    config.Format = get_file_format(input_args.input_table_file)

    if input_args.VERBOSE :
        print(f"  File format is {config.Format}")
        print_elapse_time(t0)
        sys.stdout.flush()

    # make sure table name is appropriate for file format
    config.table_name = check_table_name(input_args.table_name,config.Format)

    config.outfile = get_outfile_name(input_args,config)

    # construct command for C code
    config.command = make_command_sntable_dump(input_args,config)

    if input_args.VERBOSE :
        print(f"\n C-program command = \n    {config.command}")
        print_elapse_time(t0) 
        print('\n')
        sys.stdout.flush()

    istat = os.system(config.command)

    if input_args.VERBOSE:
        print_elapse_time(t0) 

    if istat != 0 : sys.exit(istat)

    # check option to combine_fitres files
    if input_args.append_file : 
        append_fitres(input_args,config)

    sys.exit(0)

    # === END ===
